pydantic_ai_graph/pydantic_ai_graph/state.py

```diff
@@ -19,33 +24,40 @@ def serialize(self) -> bytes | None:
         raise NotImplementedError
 
 
+RunEndT = TypeVar('RunEndT', default=None)
+NodeRunEndT = TypeVar('NodeRunEndT', covariant=True, default=Never)
 StateT = TypeVar('StateT', bound=Union[None, AbstractState], default=None)
 
 
 @dataclass
-class Snapshot:
-    """Snapshot of a graph."""
-
-    last_node_id: str
-    next_node_id: str
-    start_ts: datetime
-    duration: float
-    state: bytes | None = None
-
-    @classmethod
-    def from_state(
-        cls, last_node_id: str, next_node_id: str, start_ts: datetime, duration: float, state: StateT
-    ) -> Snapshot:
-        return cls(
-            last_node_id=last_node_id,
-            next_node_id=next_node_id,
-            start_ts=start_ts,
-            duration=duration,
-            state=state.serialize() if state is not None else None,
-        )
-
-    def summary(self) -> str:
-        s = f'{self.last_node_id} -> {self.next_node_id}'
-        if self.duration > 1e-5:
-            s += f' ({self.duration:.6f}s)'
-        return s
+class Step(Generic[StateT, RunEndT]):
+    """History item describing the execution of a step of a graph."""
+
+    state: StateT
+    node: BaseNode[StateT, RunEndT]
+    start_ts: datetime = field(default_factory=lambda: datetime.now(tz=timezone.utc))
+    duration: float | None = None
+
+    kind: Literal['start_step'] = 'start_step'
+
+    def __post_init__(self):
+        # Copy the state to prevent it from being modified by other code
+        self.state = copy.deepcopy(self.state)
+
+
+@dataclass
+class EndEvent(Generic[StateT, RunEndT]):
+    """History item describing the end of a graph run."""
+
+    state: StateT
+    result: End[RunEndT]
+    ts: datetime = field(default_factory=lambda: datetime.now(tz=timezone.utc))
+
+    kind: Literal['end'] = 'end'
+
+    def __post_init__(self):
+        # Copy the state to prevent it from being modified by other code
+        self.state = copy.deepcopy(self.state)
+
+
+GraphHistoryItem = Union[Step[StateT, RunEndT], EndEvent[StateT, RunEndT]]
```

Comment:
```suggestion
StepOrEnd = Union[Step[StateT, RunEndT], EndEvent[StateT, RunEndT]]
```
might be clearer.

Reply:
I just wasn't sure if we might add more over time, but I'm okay with that renaming if you are

Reply:
I'll rename it like this for now, if you foresee adding anything besides these though I think we should rename it now